keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,656 @@
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters. Kubeadm support is required for implementing dual-stack continuous integration (CI) tests.
+- New e2e test cases will need to be added to test parallel IPv4/IPv6 connectivity between pods, nodes, and services.
+
+### Implementation Plan
+
+Given the scope of this enhancement, it has been suggested that we break the implementation into descrete phases that may be spread out over the span of many release cycles. The phases are as follows:
+
+- pod-to-pod dual-stack routing and non-vip services
+- Single-family services
+- Multi-family services
+- External dependencies, eg. cloud-provide, CNI, CRI, CoreDNS etc...
+
+### Awareness of Multiple IPs per Pod
+
+Since Kubernetes Version 1.9, Kubernetes users have had the capability to use dual-stack-capable CNI network plugins (e.g. Bridge + Host Local, Calico, etc.), using the 
+[0.3.1 version of the CNI Networking Plugin API](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md), to configure multiple IPv4/IPv6 addresses on pods. However, Kubernetes currently captures and uses only IP address from the pod's main interface.
+
+This proposal aims to extend the Kubernetes Pod Status and the Pod Network Status API so that Kubernetes can track and make use of one or many IPv4 addresses and one or many IPv6 address assignment per pod.
+
+CNI provides list of ips and their versions. Kubernetes currently just chooses to ignore this and use a single IP. This means this struct will need to follow the pod.Pod ip style of primary IP as is, and another slice of IPs, having Pod.IPs[0] == Pod.IP which will look like the following:
+
+    type PodNetworkStatus struct {
+      metav1.TypeMeta `json:",inline"`
+
+      // IP is the primary ipv4/ipv6 address of the pod. Among other things it is the address that -
+      //   - kube expects to be reachable across the cluster
+      //   - service endpoints are constructed with
+      //   - will be reported in the PodStatus.PodIP field (will override the IP reported by docker)
+      IP net.IP `json:"ip" description:"Primary IP address of the pod"`
+            IPs  []net.IP `json:"ips" description:"list of ips"`
+    }
+
+CNI as of today does not provide additional metadata to the IP. So this Properties field - speced in this KEP - will be empty until CNI spec includes properties. Although in theory we can copy the interface name, gateway etc. into this Properties map.
+
+### Required changes to Container Runtime Interface (CRI)
```

Comment:
We may also need to update `UpdateRuntimeConfig`. For some CNI plugins, e.g. kubenet, we use that function to pass in the node CIDR.

Reply:
Agreed. Thanks